concurrency/smallTest/SmallAtomViorOrOverSync.java

```diff
@@ -0,0 +1,190 @@
+    
+    public void atomViol05() {
+        if (myMap.get(key) != null) {
+            String myVal = myMap.get(key);
+            System.out.println(myVal.length()); // ensure accesses null
+        }
+    }
+    
+    public void atomViol06() {
+        if ( myMap.containsKey(key)) {
+            String myVal = myMap.get(key);
+            System.out.println(myVal.length()); // ensure accesses null
+        }
+    }
+    
+    // ================ check and remove ===============
+    
+    public void atomViol_checkRemove_01() {
+        if (myMap.get(key) != null) {
+            String myVal = myMap.remove(key);
+            System.out.println(myVal.size());            
+        }
+    }
+    
+    // ========= check and put with additional info about the put =========
+    public void atomViol07() {
+        String myVal = myMap.get(key);
+        if (myVal == null) {
+            myMap.put(key, val);
+        }
+        return val;
+    }
+    
+    public void atomViol08() {
+        if (myMap.get(key) == null) {
+            myMap.put(key, val);
+        }
+        System.out.println(val);
+    }
+    
+    public void atomViol09() {
+        if ( ! myMap.containsKey(key)) {
+            myMap.put(key, val);
+        }
+        System.out.println(val);
+        return val;
+    }
+    
+    // ================== over sync with put ================
+ 
+    public void atomViol07() {
+        String myKey = "lala" + 88;
+        String sillyVal = "lala" + 99;
+        synchronized(this) {
+            String localVal = myMap.get(myKey);
+            if (localVal == null) {
+                myMap.put(myKey, sillyVal);
+            }
+            return myKey;
+        }
+    }
+    
+    public synchronized void atomViol08() {
+        String myKey = "lala" + 88;
+        String sillyVal = "lala" + 99;
+        if (myMap.get(myKey) == null) {
+            myMap.put(myKey, sillyVal);
+        }
+    }
+    
+    public String atomViol09() {
+        String myKey = "lala" + 88;
+        String sillyVal = "lala" + 99;
+        synchronized (this) {
+            if ( ! myMap.containsKey(myKey)) {
+                myMap.put(myKey, sillyVal);
+            }
+        }
+        System.out.println(sillyVal);
+        return sillyVal;
+    }
+    public String atomViol10() {
+        String myKey = "lala" + 88;
+        String sillyVal = "lala" + 99;
+        synchronized (this) {
+            boolean contains = myMap.containsKey(myKey);
+            if ( ! contains) {
+                myMap.put(myKey, sillyVal);
+            }
+        }
+        System.out.println(sillyVal);
+        return sillyVal;
+    }
+    
+    // =================== over sync with remove =============
+    
+    public void overSync11() {
+        String myKey = "lala" + 88;
+        String sillyVal = "lala" + 99;
+        synchronized (this) {
+            boolean contains = myMap.containsKey(myKey);
+            if ( ! contains) {
+                myMap.remove(myKey);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 These lines use `ConcurrentHashMap`&#39;s **sequential** API and are protected by `synchronized` to ensure they execute together: **148 and 150**. This synchronization may reduce performance by limiting parallelism and does not take advantage of `ConcurrentHashMap`&#39;s **concurrent** API.

**Fix**  
 Consider replacing the check and `remove(key)` with `remove(key, value)` and then removing the synchronization. If you do not need the check, consider removing the check and the synchronization and just calling `remove(key)`.

**More info**  
 `remove(key, value)` removes the key only if the current value is equal to the `value` parameter. `remove(key, value)` returns `true` if the call removed the `key` and `value`.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/4d3c580b03475a6ec9323d11e6875c542f8e3f6d/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Client.java#L1591-L1592) (external link).